aea/registries/base.py

```diff
@@ -225,292 +237,165 @@ def _add_protocol(self, protocol_directory: Path):
+        """Unregister all the components by skill."""
+        self._items.pop(skill_id, None)
+
     def setup(self) -> None:
         """
-        Set up the behaviours in the registry.
+        Set up the items in the registry.
 
         :return: None
         """
-        for behaviours in self._behaviours.values():
-            for behaviour in behaviours:
-                behaviour.setup()
+        for item in self.fetch_all():
+            item.setup()
 
     def teardown(self) -> None:
         """
         Teardown the registry.
 
         :return: None
         """
-        for skill_id, behaviours in self._behaviours.items():
-            for behaviour in behaviours:
+        for skill_id, items in self._items.items():
+            for _, item in items.items():
                 try:
-                    behaviour.teardown()
+                    item.teardown()
                 except Exception as e:
                     logger.warning(
-                        "An error occurred while tearing down behaviour {}/{}: {}".format(
-                            skill_id, type(behaviour).__name__, str(e)
+                        "An error occurred while tearing down item {}/{}: {}".format(
+                            skill_id, type(item).__name__, str(e)
                         )
                     )
-        self._behaviours = {}
 
 
-class TaskRegistry(Registry):
-    """This class implements the task registry."""
+class HandlerRegistry(ComponentRegistry[Handler]):
+    """This class implements the handlers registry."""
 
     def __init__(self) -> None:
         """
         Instantiate the registry.
 
         :return: None
         """
-        self._tasks = {}  # type: Dict[SkillId, List[Task]]
+        super().__init__()
 
-    def register(self, ids: Tuple[None, SkillId], tasks: List[Task]) -> None:
+    def fetch_by_protocol(self, protocol_id: ProtocolId) -> List[Handler]:
         """
-        Register a task.
+        Fetch the handler by the pair protocol id and skill id.
 
-        :param ids: the skill id.
-        :param tasks: the tasks list.
-        :return: None
+        :param protocol_id: the protocol id
+        :return: the handlers registered for the protocol_id and skill_id
         """
-        skill_id = ids[1]
-        if skill_id in self._tasks.keys():
-            logger.warning(
-                "Tasks already registered with skill id '{}'".format(skill_id)
+        # TODO this could be optimized by having an index by protocol
+        #      however that requires extending the superclass methods.
+        handlers = self.fetch_all()
+        return [
+            handler
+            for handler in filter(
+                lambda handler: handler.SUPPORTED_PROTOCOL == protocol_id, handlers
             )
-        self._tasks.setdefault(skill_id, []).extend(tasks)
+        ]
 
-    def unregister(self, skill_id: SkillId) -> None:
+    def fetch_by_protocol_and_skill(
```

Comment:
notice: the pair `ProtocolId, SkillId` is a unique identifier for `Handler`. So not only `SkillId, HandlerName`.

Probably it must be enforced at `register` time, and not at `fetch` time (see the implementation)... thoughts?

Reply:
Yes I think it should be enforced at register time. Is it worth keeping an assertion at fetch time though?